indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -828,34 +911,44 @@ public TaskStatusPlus apply(TaskRunnerWorkItem workItem)
     }
   }
 
-  private Collection<? extends TaskRunnerWorkItem> securedTaskRunnerWorkItem(
-      Collection<? extends TaskRunnerWorkItem> collectionToFilter,
+  private List<TaskStatusPlus> securedTaskStatusPlus(
+      List<TaskStatusPlus> collectionToFilter,
+      Map<String, String> taskInfoCache,
+      @Nullable String dataSource,
+      @Nullable String type,
       HttpServletRequest req
   )
   {
-    Function<TaskRunnerWorkItem, Iterable<ResourceAction>> raGenerator = taskRunnerWorkItem -> {
-      final String taskId = taskRunnerWorkItem.getTaskId();
-      final Optional<Task> optionalTask = taskStorageQueryAdapter.getTask(taskId);
-      if (!optionalTask.isPresent()) {
+    Function<TaskStatusPlus, Iterable<ResourceAction>> raGenerator = taskStatusPlus -> {
+      final String taskId = taskStatusPlus.getId();
+      String taskDatasource = taskInfoCache.get(taskId);
+      if (taskDatasource == null) {
         throw new WebApplicationException(
             Response.serverError().entity(
                 StringUtils.format("No task information found for task with id: [%s]", taskId)
             ).build()
         );
       }
-
       return Lists.newArrayList(
           new ResourceAction(
-              new Resource(optionalTask.get().getDataSource(), ResourceType.DATASOURCE),
+              new Resource(taskDatasource, ResourceType.DATASOURCE),
               Action.READ
           )
       );
     };
-
+    List<TaskStatusPlus> dataSourceFilter = collectionToFilter;
```

Comment:
For the case where the caller provides a datasource, I think `getTasks` could be optimized in a couple of ways:
- We could offload filtering of task by datasource to the metadata store, this would cut down on the amount of data being transferred to and processed on the overlord and also allow users to set up indices for filtering by datasource on the task table if needed
- If we do that, we could additionally reduce the number of authorization checks needed from (number of tasks) to one (just check that that user has access to the datasource being filtered on once)

Reply:
yes, we could add datasource as an argument to those methods to put in a "WHERE datasource = x" clause to the SQL queries being made

Reply:
ok, added `WHERE datasource = :ds` clause to all sql  impementations